### PR TITLE
Account Transaction : Register Data

### PR DESCRIPTION
**Pull request by romil-stactrace**
## Purpose

Support for Account Transaction Register Data

## Changes
Added RegisterData Transaction

## Checklist
- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x]  (If necessary) I have updated the CHANGELOG.

## CLA acceptance
By submitting the contribution I accept the terms and conditions of the
Contributor License Agreement v1.0
- link: https://developers.concordium.com/CLAs/Contributor-License-Agreement-v1.0.pdf

- [x] I accept the above linked CLA.

### CODE REVIEW

**Review by parv0888 — CHANGES_REQUESTED** [review]

File: concordium-sdk/pom.xml [hunk 1]
```diff
@@ -195,4 +195,4 @@
             </plugins>
         </pluginManagement>
     </build>
-</project>
+</project>
```
Comment (parv0888) [inline concordium-sdk/pom.xml:198]: revert this change

**Review by parv0888 — CHANGES_REQUESTED** [review]

**Comment by parv0888** [comment]
@romil-stactrace  - please update ChangeLog

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/Transaction.java [hunk 2]
```diff
@@ -51,4 +51,23 @@ static void verifyTransferInput(AccountAddress sender, AccountNonce nonce, Expir
             throw TransactionCreationException.from(new IllegalArgumentException("Signer cannot be null or empty"));
         }
     }
+    
+    static void verifyRegisterDataInput(AccountAddress sender, AccountNonce nonce, Expiry expiry, Data data, TransactionSigner signer) throws TransactionCreationException {
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/transactions/Transaction.java:55]: I know that `verifyTransferInput` is currently present in `Transaction.java` however this is for historic reasons as there were only `Transfer` and `TransferWithMemo` before.

I think now is the time to factor these verification functions away from `Transaction.java` as it will quickly ramp up.

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/Data.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import org.apache.commons.codec.binary.Hex;
+
+import java.nio.ByteBuffer;
+import java.util.Objects;
+
+@EqualsAndHashCode
+@ToString
+public class Data {
+
+    public final static int BYTES = 256;
```
[the remark this answers is not in the document]
Comment (EmilLa1) [reply concordium-sdk/src/main/java/com/concordium/sdk/transactions/Data.java:19]: Max size can be seen here: 
https://github.com/Concordium/concordium-base/blob/1cb35e6567a8d9cf72b2934e94a16625a62ffb61/haskell-src/Concordium/Types.hs#L811

**Review by EmilLa1 — CHANGES_REQUESTED** [review]
Overall looks good, I only have a few minor comments.  :+1:

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/Data.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+import lombok.val;
+import org.apache.commons.codec.DecoderException;
+import org.apache.commons.codec.binary.Hex;
+
+import java.nio.ByteBuffer;
+import java.util.Objects;
+
+@EqualsAndHashCode
+@ToString
+public class Data {
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/transactions/Data.java:20]: added java docs

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/Data.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/transactions/Data.java:25]: Added check for Max Size

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/Data.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    // return the length of the serialized data
+    int getLength() {
+        return UInt16.BYTES + value.length;
+    }
+
+    static Data from(byte[] data) throws TransactionCreationException {
+        if (Objects.isNull(data)) {
+            throw TransactionCreationException.from(new IllegalArgumentException("Data cannot be null"));
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/transactions/Data.java:89]: Added check for max size

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/Transaction.java (hunk 2 above)
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/transactions/Transaction.java]: Refactored and moved the method to Builder.

**Review by EmilLa1 — APPROVED** [review]
lgtm 👍 Good job

**Review by EmilLa1 — CHANGES_REQUESTED** [review]
Still looks good :) 
But could you please add it to: 

https://github.com/Concordium/concordium-java-sdk/blob/main/README.md#supported-transactions

File: README.md [hunk 6]
```diff
@@ -205,6 +205,8 @@ Sends funds from one account to another.
 - TransferWithMemo
 Sends funds from one account to another with an associated `Memo`.
 
+- Register Data
+Sends arbitary byte data to be registered on chain.
```
Comment (EmilLa1) [inline README.md:209]:
```suggestion
Registers a maximum of 256 bytes on the chain.
```

**Review by EmilLa1 — CHANGES_REQUESTED** [review]
Looks good 👍 
I think we should just rephrase it a bit.

**Review by EmilLa1 — APPROVED** [review]
Looks good 👍
